jax/experimental/rnn.py

```diff
@@ -308,12 +317,15 @@ def lstm_cell(carry, x, *, W_ih, W_hh, b_ih, b_hh):
   for l in range(num_layers * 2):
     cell = partial(
         lstm_cell, W_ih=W_ih[l], W_hh=W_hh[l], b_ih=b_ih[l], b_hh=b_hh[l])
+    cell_fn = partial(scan_fn, cell)
     if l % 2 == 0:
-      (h_t, c_t), seq_first_y_fwd = jax.lax.scan(cell, (h_0[l], c_0[l]),
+      out = jax.lax.scan(cell_fn, (h_0[l], c_0[l]),
                                                  seq_first_y)
+      (h_t, c_t), seq_first_y_fwd = _extract_output(seq_lengths, out)
     else:
-      (h_t, c_t), seq_first_y_bwd = jax.lax.scan(
-          cell, (h_0[l], c_0[l]), seq_first_y, reverse=True)
+      out = jax.lax.scan(
```

Comment:
I think there's an issue here.

The way that it's set up, the `h, c` are returned as part of the scan `ys`. However, since the `scan` has `reverse=True`, the `seq_first_y` will be in the right order but the `h_t, c_t` will be reversed.

See this example:
```python
def body(i, x):                            
  return i + 1, (i, x + 1)                 
                                           
print(jax.lax.scan(body, 0, jnp.arange(4), reverse=True))
# Prints (4, ([3, 2, 1, 0], [1, 2, 3, 4]))
```

See how the `i` in the carry was reversed but the `x + 1` is not.

Therefore when you call `_extract_output` on `h_t, c_t`, it will be selecting out the *initial* state, not the *last* state. 

There are two options for a fix here:

1) reverse `h_t, c_t` after they are returned from the scan
2) instead of using `scan(..., reverse=True)`, manually reverse `seq_first_y` and reverse `seq_first_y_bwd` at the end.

Reply:
You are totally right @sharadmv! `2` is the correct approach, however, you cannot just revert, you have to revert while keeping padding at the end. I just pushed some logic we had for this.

Reply:
Check out the `_flip_sequence` function.